---
title: "Dealing with Correlated High-Dimensional Data"
author: "Gokhan Ciflikli"
date: '2017-10-11'
output:
  html_document:
    df_print: paged
  blogdown::html_page:
    fig_retina: 2
    fig_width: 7
description: Comparison of random forests vs. fuzzy forests when *p >> n*.
slug: fuzzy-forest
tags:
- fuzzyforest
- R
- randomForest
categories:
- R
- Prediction
- Visualization
---



<div id="theory-to-application-1" class="section level3">
<h3>Theory to Application <a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a></h3>
<p>I recently realized my review of Michael Alvarez’s edited volume <a href="http://journals.sagepub.com/doi/abs/10.1177/1478929917712143">Computational Social Science: Discovery and Prediction</a> went online a while ago. If memory serves, used to be the case that book reviews were freely available; alas now even a 400-word long piece is behind a paywall. I initially planned to do a post covering the whole book—at least the application part, as the volume is neatly segregated on theory vs. application lines. And instead of going through various applied tools, which is what the edited volume successfully accomplishes, I decided to pick a particular method and give it a test drive. Disclaimer—the example would not necessarily pass as a social science enterprise, but the data challenge it poses makes it relevant for all type of users.</p>
<p><a href="http://escholarship.org/uc/item/55h4h0w7">Fuzzy forests</a>, in addition to being situated on the better side of the statistical method naming spectrum (<em>don’t</em> you just want to use them?), advertises a neat improvement over the regular <code>randomForest</code> package in R—the ability to deal with correlated, high dimensional data. Cast in the fiery realms of biostats, the algorithm relies on Weighted Gene Coexpression Network Analysis (WGCNA) to create uncorrelated sets of features called <em>modules</em>. These modules in turn are separately fed to recursive feature elimination random forests (RFE-RFs). The ‘surviving’ features—the ones that made it to the last round—go through a final round of RFE-RFs. At this point, based on a user-specified length of a ranked variable importance list (more on that in a bit), the selected features can be used to train a predictive model.</p>
<p>Given this high dimensional premise, let’s compare the vanilla random forest algorithm to the fuzzy forest. If fuzzy forests sound like fitting a random forest after doing a principal component analysis, you are not alone. However, this inevitably invites questions about interpretability: Random forests get enough flak for being black boxes, and now we are adding extra <em>mathemagic</em> by introducing loadings. One might argue it is better to stick with regular random forests if the marginal utility of making them ‘fuzzy’ does not pay off.</p>
<p>The standard way of comparing these two algorithms would be simply to evalute the predictive ability of both in the presence of high dimensional data. However, this is more or less the point of the introductory paper linked earlier. Let’s consider the interpretability and stability of the results instead. Do the algorithms generate convergent or divergent insights? Fuzzy forests require preprocessing—is it worth the trouble if one can get equivalent results from RFs out of the box? Let’s find out.</p>
</div>
<div id="data-setup" class="section level3">
<h3>Data Setup</h3>
<p><code>fuzzyforest</code> comes prepackaged with <a href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/">a dataset</a> on gene expression levels in liver tissue. The attached subset has 66 observations (female mice) and 3600 variables (gene expression measurements); the good old 1:600 ratio of <em>p</em> to <em>n</em>. The first variable in the dataset is weight (g), which I will treat as the outcome we want to predict. Unsurprisingly, its a continuous variable. Later, given 1:600 <em>p</em> to <em>n</em> ratio is not serious enough of a handicap with an <em>n</em> of 66, we will go ahead and shoot ourselves in the <em>other</em> foot (lose more data) by converting the outcome into a binary variable. Although this is probably a pretty terrible idea, this superficial constraint might shed light on the extent of the algorithm capabilities (i.e. robustness or its lack of) in our comparison.</p>
<pre class="r"><code>library(fuzzyforest)
data(Liver_Expr)
dim(Liver_Expr) #yikes</code></pre>
<pre><code>## [1]   66 3601</code></pre>
<pre class="r"><code>weight &lt;- Liver_Expr[, 1]
expression_levels &lt;- Liver_Expr[, -1]
summary(weight)</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   22.20   33.70   37.10   37.25   42.45   54.10</code></pre>
<pre class="r"><code>hist(weight) #look at that perfectly coincidental normal distribution</code></pre>
<p><img src="/post/2017-10-11-fuzzy-forest_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
</div>
<div id="random-forest" class="section level3">
<h3>Random Forest</h3>
<p>Let’s take a step back and try fitting a random forest without going through the WCGNA process. For the first pass, we will not bother tuning the parameters—we just want to obtain a ballpark OOB error rate to act as a baseline. Note that you can call <code>tuneRF</code> to get optimal ntree and mtry values.</p>
<pre class="r"><code>library(randomForest)
set.seed(1895)
rf_fit &lt;- randomForest(expression_levels, weight, ntree = 1000, importance = TRUE)
rf_fit</code></pre>
<pre><code>## 
## Call:
##  randomForest(x = expression_levels, y = weight, ntree = 1000,      importance = TRUE) 
##                Type of random forest: regression
##                      Number of trees: 1000
## No. of variables tried at each split: 1200
## 
##           Mean of squared residuals: 12.90961
##                     % Var explained: 71.92</code></pre>
<pre class="r"><code>varImpPlot(rf_fit, n.var = 10)</code></pre>
<p><img src="/post/2017-10-11-fuzzy-forest_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>Even though we have 1000 trees for 3600 variables, the RF algorithm considers 1200(!) variables to choose from that pool for every decision for regression problems by default. Our casual model explains around 72% of the variance, and we plotted the top ten predictors based on their performance. If you are particularly lazy, you can extract them using the <code>pdp</code> package:</p>
<pre class="r"><code>library(pdp)
topPredictors(rf_fit, n = 10)</code></pre>
<pre><code>##  [1] &quot;MMT00019254&quot; &quot;MMT00067823&quot; &quot;MMT00072042&quot; &quot;MMT00002575&quot; &quot;MMT00038934&quot;
##  [6] &quot;MMT00063359&quot; &quot;MMT00067105&quot; &quot;MMT00030931&quot; &quot;MMT00028861&quot; &quot;MMT00033001&quot;</code></pre>
<p>This was for regression. Let’s have a quick glance at its classification variant:</p>
<pre class="r"><code>Liver_Expr$weight_class &lt;- ifelse(Liver_Expr$weight &lt; mean(Liver_Expr$weight), 0, 1)
weight_class &lt;- as.factor(Liver_Expr$weight_class)
rf_fit_cl &lt;- randomForest(expression_levels, weight_class, ntree = 1000, importance = TRUE)
rf_fit_cl</code></pre>
<pre><code>## 
## Call:
##  randomForest(x = expression_levels, y = weight_class, ntree = 1000,      importance = TRUE) 
##                Type of random forest: classification
##                      Number of trees: 1000
## No. of variables tried at each split: 60
## 
##         OOB estimate of  error rate: 12.12%
## Confusion matrix:
##    0  1 class.error
## 0 32  3  0.08571429
## 1  5 26  0.16129032</code></pre>
<pre class="r"><code>varImpPlot(rf_fit_cl, n.var = 10)</code></pre>
<p><img src="/post/2017-10-11-fuzzy-forest_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Out of the box, we get an OOB estimate of 12%. This time, the default choice of mtry is 60 (down from 1200 for the regression problem). We see a slight tendency for the ‘lighter’ class—the RF did a better job identifying the zeroes. In general, we find there is quite the overlap between the classification and regression results in terms of the top ten most important variables. However, this is not necessarily a smoking gun; the heavier weight <em>class</em> might have a slightly different set of characteristics compared to that of a <em>continuum</em> of weights. On a final note, you can generate a full report of your forest via the aptly named <code>randomForestExplainer</code> package. To illustrate, calling <code>explain_forest(rf_fit)</code> would result in <a href="\post\2017-10-11-fuzzy-forest_files\figure-html\forest.html">this document</a>. Definitely not a black box anymore!</p>
</div>
<div id="fuzzy-forest" class="section level3">
<h3>Fuzzy Forest</h3>
<p>As I mentioned earlier, <code>fuzzyforest</code> depends on WGCNA for dimension reduction—we cannot fit a model right away. Although it is on CRAN, at least one dependency is on Bioconducter. To install <code>WGCNA</code>, run:</p>
<pre class="r"><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;AnnotationDbi&quot;, &quot;impute&quot;, &quot;GO.db&quot;, &quot;preprocessCore&quot;))
install.packages(&quot;WGCNA&quot;)</code></pre>
<p>After you load the library, you can follow the example code from the publication to fit the model quickly. Note that this is somewhat optimized; at least in comparison to the RF setup we have above.</p>
<pre class="r"><code>WGCNA_params &lt;- WGCNA_control(power = 6, 
                              TOMType = &quot;unsigned&quot;, 
                              minModuleSize = 30,
                              numericLabels = TRUE, 
                              pamRespectsDendro = FALSE)
mtry_factor &lt;- 1; drop_fraction &lt;- .25; number_selected &lt;- 10
keep_fraction &lt;- .05; min_ntree &lt;- 5000; ntree_factor &lt;- 5
final_ntree &lt;- 5000;
screen_params &lt;- screen_control(drop_fraction = drop_fraction,
                                keep_fraction = keep_fraction,
                                min_ntree = min_ntree,
                                mtry_factor = mtry_factor,
                                ntree_factor = ntree_factor)
select_params &lt;- select_control(drop_fraction = drop_fraction,
                                number_selected = number_selected,
                                min_ntree = min_ntree,
                                mtry_factor = mtry_factor,
                                ntree_factor = ntree_factor)</code></pre>
<p>I will not pretend Rmarkdown can handle fitting 3600 variables on the fly for the <code>wff_fit</code> object (the page would have taken a while to load). Hence, the below code is not evaluated. When you run it, don’t forget to change the number of processors at the end depending on your setup. The WGCNA package will produce a warning with instructions if it detects a multi-core system but no parallel processing enabled.</p>
<pre class="r"><code>wff_fit &lt;- wff(expression_levels, weight,
               WGCNA_params = WGCNA_params,
               screen_params = screen_params,
               select_params = select_params,
               final_ntree = final_ntree,
               num_processors = 8)</code></pre>
<p>Once you are done fitting, you can access the results in several ways. You can get a list of the top predictors by calling <code>print(wff_fit)</code>.<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a> Finally, you can obtain a nice in-house graph plotting the module distribution by typing <code>modplot(wff_fit)</code>. This gives us the relative importance of the modules. You should get something like the following:</p>
<div class="figure">
<img src="/img/module.png" />

</div>
<div class="figure">
<img src="/img/varimp.png" />

</div>
<p>And for the classification problem:</p>
<div class="figure">
<img src="/img/varimp2.png" />

</div>
<p>Again, somewhat similar variable selection, however here we see more variation. On the other hand, the similarities are quite striking when present: the top predictor (<em>expression 19854</em>) is the same, and modules 7 and 11 are the most influential ‘loadings’ for both types of problems.</p>
<p>If you are curious about how the fuzzy forest modules compare to PCA loadings, you can get an idea by setting the component number manually so that it matches the WGCNA process (22 in our example):</p>
<pre class="r"><code>library(pcaMethods)
PCAmethods &lt;- pca(Liver_Expr[, -1],
                  scale = &quot;uv&quot;,
                  center = T,
                  nPcs = 22,
                  method = &quot;svd&quot;)
print(PCAmethods)</code></pre>
<pre><code>## svd calculated PCA
## Importance of component(s):
##                  PC1    PC2    PC3     PC4     PC5    PC6     PC7     PC8
## R2            0.1883 0.1441 0.1215 0.06953 0.03955 0.0381 0.03553 0.03134
## Cumulative R2 0.1883 0.3323 0.4538 0.52332 0.56287 0.6010 0.63650 0.66784
##                   PC9    PC10    PC11    PC12    PC13    PC14    PC15
## R2            0.02848 0.02422 0.02009 0.01973 0.01678 0.01624 0.01533
## Cumulative R2 0.69632 0.72054 0.74063 0.76036 0.77714 0.79338 0.80871
##                  PC16   PC17   PC18    PC19    PC20    PC21    PC22
## R2            0.01403 0.0128 0.0124 0.01146 0.00974 0.00919 0.00865
## Cumulative R2 0.82274 0.8355 0.8479 0.85940 0.86914 0.87833 0.88698
## 3601     Variables
## 66   Samples
## 0    NAs ( 0 %)
## 22   Calculated component(s)
## Data was mean centered before running PCA 
## Data was scaled before running PCA 
## Scores structure:
## [1] 66 22
## Loadings structure:
## [1] 3601   22</code></pre>
<p>We see that the increase in <span class="math inline">\(R^2\)</span> begins to stagnate after the third component, a narrative we can corroborate with the FF modules for the regression problem (without making a claim about the makeup of the clusters). One can feed the loadings into a RF, however that will probably increase the difficulty level of your presentation if you actually need to convey your findings to a third-party. This is what it comes down to with FF as well.</p>
<p>Verdict? It’s a hard sell (for social science purposes).<a href="#fn3" class="footnoteRef" id="fnref3"><sup>3</sup></a> The performance of the RF—in terms of feature seletion—out of the box with no tuning looks like on par with what FF has to offer. Plus, FF requires preprocessing, either via <code>WGCNA</code> or prior knowledge, to work in the first place. The wide range of R packages that accept <code>randomForest</code> output is another decisive advantage. Don’t abandon the RF ship just yet.</p>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>My attempts at finding a cheesy title were discouraged by multiple articles sharing the title <em>A Tale of Two Forests</em> <a href="https://www.jstor.org/stable/3034035?seq=1#page_scan_tab_contents">1</a> <a href="http://agris.fao.org/agris-search/search.do?recordID=US201600103710">2</a> <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0085993">3</a> <a href="https://ehp.niehs.nih.gov/121-a78/">4</a>.<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p>The vignette claims you can call <code>varImpPlot</code> from <code>randomForest</code> to evaluate the <code>wff_fit</code> object, however I get an error with my current setup (“This function only works for objects of class randomForest”).<a href="#fnref2">↩</a></p></li>
<li id="fn3"><p>Naturally, in a tongue-in-cheek way. Kudos to the package developers for investing their time and creating cool things for us to play with.<a href="#fnref3">↩</a></p></li>
</ol>
</div>
